File: fib.go
// go run fib.go
// Fib ElapsedTime in seconds: 2.518877625s
package main

import (
    "fmt"
    "time"
    "log"
)

func fib(n int) int {
    if n <= 0 {
        return 0
    }
    if n == 1 || n == 2 {
        return 1
    }
    return fib(n-1) + fib(n-2)
}

func trace(s string) (string, time.Time) {
    log.Println("START:", s)
    return s, time.Now()
}

func un(s string, startTime time.Time) {
    endTime := time.Now()
    log.Println("  END:", s, "ElapsedTime in seconds:", endTime.Sub(startTime))
}

func main() {
    defer un(trace("Fib"))
    for i := 1; i <= 42; i++ {
        fmt.Printf("%d ", fib(i))
    }
    fmt.Println()
}

File: fib.rb
# ruby fib.rb - 75.544086 seconds

def fib(n)
    if n <= 0
        return 1
    end
    if n == 1 or n == 2
        return 1
    end
    return fib(n-1) + fib(n-2)
end

def main()
    ntimes = 42
    t1 = Time.now

    for i in 1..ntimes
        print "%s " % fib(i)
    end

    puts ""
    t2 = Time.now
    delta = t2 - t1
    puts "%f seconds" % delta
end

main()
